### PR TITLE
support imagePullSecrets and imagePullPolicy in kubefed init

### LINKED ISSUE #50718: Enable `kubefed init` support `ImagePullSecrets` and `imagePullPolicy`
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
> /kind bug

/kind feature

**What happened**:
The `kubefed init` can only specify the image but cannot specify the `ImagePullSecrets` and `imagePullPolicy`, this will cause issues if I want to use images from private registry.
```
      --api-server-advertise-address string      Preferred address to advertise api server nodeport service. Valid only if 'api-server-service-type=NodePort'.
      --api-server-service-type string           The type of service to create for federation API server. Options: 'LoadBalancer' (default), 'NodePort'. (default "LoadBalancer")
      --apiserver-arg-overrides string           comma separated list of federation-apiserver arguments to override: Example "--arg1=value1,--arg2=value2..."
      --apiserver-enable-basic-auth              Enables HTTP Basic authentication for the federation-apiserver. Defaults to false.
      --apiserver-enable-token-auth              Enables token authentication for the federation-apiserver. Defaults to false.
      --controllermanager-arg-overrides string   comma separated list of federation-controller-manager arguments to override: Example "--arg1=value1,--arg2=value2..."
      --dns-provider string                      Dns provider to be used for this deployment.
      --dns-provider-config string               Config file path on local file system for configuring DNS provider.
      --dns-zone-name string                     DNS suffix for this federation. Federated Service DNS names are published with this suffix.
      --dry-run                                  dry run without sending commands to server.
      --etcd-image string                        Image to use for etcd server. (default "gcr.io/google_containers/etcd:3.0.17")
      --etcd-persistent-storage                  Use persistent volume for etcd. Defaults to 'true'. (default true)
      --etcd-pv-capacity string                  Size of persistent volume claim to be used for etcd. (default "10Gi")
      --etcd-pv-storage-class string             The storage class of the persistent volume claim used for etcd.   Must be provided if a default storage class is not enabled for the host cluster.
      --federation-system-namespace string       Namespace in the host cluster where the federation system components are installed (default "federation-system")
      --host-cluster-context string              Host cluster context
      --image string                             Image to use for federation API server and controller manager binaries. (default "gcr.io/google_containers/hyperkube-amd64:v0.0.0-master+$Format:%h$")
      --kubeconfig string                        Path to the kubeconfig file to use for CLI requests.
```

**What you expected to happen**:
I can set `ImagePullSecrets` and `imagePullPolicy` for `kubefed init`.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/sig federation
/cc @kubernetes/sig-federation-feature-requests

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #50718

**Special notes for your reviewer**:
/assign @gyliu513 

**Release note**:

```release-note
support imagePullSecrets and imagePullPolicy in kubefed init
```

### CODE REVIEW

File: federation/pkg/kubefed/init/init.go [hunk 1]
```diff
@@ -159,6 +161,8 @@ type initFederationOptions struct {
 func (o *initFederationOptions) Bind(flags *pflag.FlagSet, defaultServerImage, defaultEtcdImage string) {
 	flags.StringVar(&o.dnsZoneName, "dns-zone-name", "", "DNS suffix for this federation. Federated Service DNS names are published with this suffix.")
 	flags.StringVar(&o.serverImage, "image", defaultServerImage, "Image to use for federation API server and controller manager binaries.")
+	flags.StringVar(&o.imagePullPolicy, "imagePullPolicy", string(api.PullIfNotPresent), "PullPolicy describes a policy for if/when to pull a container image. The default pull policy is IfNotPresent which will not pull an image if it already exists.")
```
Comment (gyliu513) [inline federation/pkg/kubefed/init/init.go]: I'd prefer we use `image-pull-policy` as the flag

File: federation/pkg/kubefed/init/init.go [hunk 2]
```diff
@@ -159,6 +161,8 @@ type initFederationOptions struct {
...
+	flags.StringVar(&o.imagePullSecrets, "imagePullSecrets", "", "Provide secrets that can access the private registry.")
```
Comment (gyliu513) [inline federation/pkg/kubefed/init/init.go]: I'd prefer we use `image-pull-secrets` as the flag

File: federation/pkg/kubefed/init/init.go [hunk 3]
```diff
@@ -361,7 +365,7 @@ func (i *initFederation) Run(cmdOut io.Writer, config util.AdminConfig) error {
 
 	fmt.Fprint(cmdOut, "Creating federation component deployments...")
 	glog.V(4).Info("Creating federation control plane components")
-	_, err = createAPIServer(hostClientset, i.commonOptions.FederationSystemNamespace, serverName, i.commonOptions.Name, i.options.serverImage, i.options.etcdImage, advertiseAddress, serverCredName, i.options.apiServerEnableHTTPBasicAuth, i.options.apiServerEnableTokenAuth, i.options.apiServerOverrides, pvc, i.options.dryRun)
+	_, err = createAPIServer(hostClientset, i.commonOptions.FederationSystemNamespace, serverName, i.commonOptions.Name, i.options.serverImage, i.options.imagePullPolicy, i.options.imagePullSecrets, i.options.etcdImage, advertiseAddress, serverCredName, i.options.apiServerEnableHTTPBasicAuth, i.options.apiServerEnableTokenAuth, i.options.apiServerOverrides, pvc, i.options.dryRun)
```
Comment (gyliu513) [inline federation/pkg/kubefed/init/init.go]: Put new parameters to the end, ditto for others

File: federation/pkg/kubefed/init/init_test.go [hunk 4]
```diff
@@ -263,6 +275,7 @@ func TestInitFederation(t *testing.T) {
 		cmd.Flags().Set("dns-zone-name", tc.dnsZoneName)
 		cmd.Flags().Set("image", tc.serverImage)
 		cmd.Flags().Set("etcd-image", tc.etcdImage)
+		cmd.Flags().Set("imagePullPolicy", tc.imagePullPolicy)
```
Comment (gyliu513) [inline federation/pkg/kubefed/init/init_test.go]: losing image pull secrets here?

File: federation/pkg/kubefed/init/init_test.go [hunk 5]
```diff
@@ -279,6 +292,9 @@ func TestInitFederation(t *testing.T) {
 		if tc.etcdPersistence != "true" {
 			cmd.Flags().Set("etcd-persistent-storage", tc.etcdPersistence)
 		}
+		if tc.imagePullSecrets != "" {
```
Comment (gyliu513) [inline federation/pkg/kubefed/init/init_test.go]: also need to check image pull policy

File: federation/pkg/kubefed/init/init.go (hunk 2 above)
Comment (dixudx) [inline federation/pkg/kubefed/init/init.go]: SGTM. Thanks.

File: federation/pkg/kubefed/init/init.go (hunk 3 above)
Comment (dixudx) [inline federation/pkg/kubefed/init/init.go]: @gyliu513 I was thinking to do like this. But current params of `createAPIServer` seem to get grouped by their type. Directly appending new params will break current style. WDYT?

File: federation/pkg/kubefed/init/init_test.go (hunk 4 above)
Comment (dixudx) [inline federation/pkg/kubefed/init/init_test.go]: adding it in below L295.

File: federation/pkg/kubefed/init/init_test.go (hunk 5 above)
Comment (dixudx) [inline federation/pkg/kubefed/init/init_test.go]: `imagePullPolicy` is set to a default value, just like `serverImage`. Still need checking?

**Comment by dixudx** [comment]
@gyliu513 Updated. PTAL. Thanks.

File: federation/pkg/kubefed/init/init_test.go [hunk 6]
```diff
@@ -621,7 +637,7 @@ func TestCertsHTTPS(t *testing.T) {
 	}
 }
 
-func fakeInitHostFactory(apiserverServiceType v1.ServiceType, federationName, namespaceName, advertiseAddress, lbIp, dnsZoneName, serverImage, etcdImage, dnsProvider, dnsProviderConfig, etcdPersistence, etcdPVCapacity, etcdPVStorageClass, apiserverOverrideArg, cmOverrideArg, tmpDirPath string, apiserverEnableHTTPBasicAuth, apiserverEnableTokenAuth, isRBACAPIAvailable bool) (cmdutil.Factory, error) {
+func fakeInitHostFactory(apiserverServiceType v1.ServiceType, federationName, namespaceName, advertiseAddress, lbIp, dnsZoneName, serverImage, imagePullPolicy, imagePullSecrets, etcdImage, dnsProvider, dnsProviderConfig, etcdPersistence, etcdPVCapacity, etcdPVStorageClass, apiserverOverrideArg, cmOverrideArg, tmpDirPath string, apiserverEnableHTTPBasicAuth, apiserverEnableTokenAuth, isRBACAPIAvailable bool) (cmdutil.Factory, error) {
```
Comment (gyliu513) [inline federation/pkg/kubefed/init/init_test.go]: put new added parameters to the end as well.

**Review by gyliu513 — COMMENTED** [review]
/lgtm

/cc @kubernetes/sig-federation-pr-reviews

**Comment by dixudx** [comment]
/assign @nikhiljindal @quinton-hoole
PTAL. Need your `/approval`. Thanks.

**Comment by dixudx** [comment]
ping @gyliu513 Rebased. Needs your `lgtm` again. Thanks.

**Comment by dixudx** [comment]
ping @irfanurrehman @gyliu513 PTAL. Needs your `lgtm`. Thanks.

**Pull request merged**